filebeat/input/v2/statestore/store.go

```diff
@@ -0,0 +1,132 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package statestore
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/registry"
+)
+
+// Store provides some coordinates access to a registry.Store.
+// All update and read operations require users to acquire an resource first.
+// A Resource must be locked before it can be modified. This ensures that at most
+// one go-routine has access to a resource. Lock/TryLock/Unlock can be used to
+// coordinate resource access even between independent components.
+type Store struct {
+	active bool
+	waiter chan struct{}
+
+	log *logp.Logger
+
+	persistentStore *registry.Store
+
+	resourcesMux sync.Mutex
+	resources    table
+}
+
+// NewStore creates a new state Store that is backed by a persistent store.
+func NewStore(log *logp.Logger, store *registry.Store) *Store {
+	invariant(log != nil, "missing a logger")
+	invariant(store != nil, "missing a persistent store")
+
+	return &Store{
+		active:          true,
+		log:             log,
+		persistentStore: store,
+		resources:       table{},
+		waiter:          make(chan struct{}),
+	}
+}
+
+// Deactivate signals the store to close itself. Resources can not be accessed anymore,
+// but in progress resource updates are still active, until they are eventually ACKed.
+// The underlying persistent store will be finally closed once all pending updates
+// have been written to the persistent store.
+func (s *Store) Deactivate() {
+	s.resourcesMux.Lock()
+	defer s.resourcesMux.Unlock()
+	s.active = false
+}
```

Comment:
Do we need to expose `Deactivate`? Can we only have `Close` that call `Deactivate`? I am asking this what is the use-case for deactivate a store because a deactivated store cannot be brought back online, unless I am mistaken?

Reply:
This is still subject to change. Idea was to keep the store alive for as long as there are still update operations in-flight. But disallow creating new resources.

Reply:
I rather introduced the concept of a session and removed Deactivate as is.